### PR TITLE
Uses Poetry for dependency management instead of having to deal with updating subdependencies within a requirements.txt file for pinned versions

### PR DESCRIPTION
**Pull request by kalaspuff**
## Poetry as dependency manager instead of `pip` + `requirements.txt`
Poetry is usually a nicer way of handling Python dependencies nowadays than having to manage `requirements.txt` like dependency files and dealing with versions of subdependencies, etc. Poetry can also handle virtualenvs automatically without additional user input if used outside of Docker environments. 

See:
* https://python-poetry.org/
* https://github.com/python-poetry/poetry

## Additional updates
Note that this PR will also update `anyio` (from 3.4) to 3.5 and `pyrsistent` (from 0.18.0) to 0.18.1.

## Poetry usage
The `poetry.lock` lockfile can be updated with the latest available versions for dependencies + subdependencies using the `poetry` CLI which will resolve which versions to use of all the used packages.

#### Lookup dependencies from `pyproject.toml` and run version resolver to update `poetry.lock`
```sh
$ poetry lock
```

#### Install the deltas from the current environment and the lock file with 
```sh
$ poetry install

# or to install packages excluding the packages listed as dev-packages (currently pytest, etc.)
$ poetry install --no-dev
```

#### List currently used dependencies
```
$ poetry show
```

#### Complete dependency tree
```
$ poetry show --tree
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,22 +33,26 @@ Otherwise, you can follow these steps:
...
+2. Install poetry (Python dependency management tooling) via pipx
    ```sh
-   pip install -r requirements.txt
+   pip install pipx && pipx install poetry
    ```
-3. Setup local .env file
+3. Install packages
    ```sh
-   copy .env.template .env
+   poetry install
```
Comment (pedropregueiro) [inline README.md:42]: if I'm not using `virtualenv` or `pyenv` (which I was before), and I run `poetry install`, will it create a venv by default and then use it for all `poetry run` stuff, is that it?

**Review by pedropregueiro — COMMENTED** [review]
this is great! left a few non-critical comments/questions, but seems good to push fwd.

can you resolve conflicts and then I'll merge later?

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,42 @@
+[tool.poetry]
+name = "app"
+version = "0.0.0"
```
[the remark this answers is not in the document]
Comment (kalaspuff) [reply pyproject.toml:3]: Preferably we would want to omit the `version` and `name` fields entirely here as they're mostly useful for libs or distributables to pypi, and usually not really used by service codes. However they are required for dependencies to be able to be installed at all. 🙃

File: README.md (hunk 1 above)
Comment (kalaspuff) [reply README.md:42]: @pedropregueiro Yes, exactly. The default is that `poetry` handles the virtualenvs on its own (unless you've already activated a virtualenv on your own). Virtualenvs via `poetry` can however be completely disabled with:
```sh
poetry config virtualenvs.create false 
poetry config virtualenvs.in-project false
```

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (kalaspuff) [reply pyproject.toml]: Added comments to explain why the fields are there at all.

**Comment by kalaspuff** [comment]
> this is great! left a few non-critical comments/questions, but seems good to push fwd.
> 
> can you resolve conflicts and then I'll merge later?

👋 Made some addition to resolve some of the ideas from the comments and resolved the conflicts from `main` branch.

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+umongo = "^3.1"
+uvicorn = "^0.16"
+uvloop = "^0.16"
+watchgod = "^0.7"
+web3 = "^5.26"
+
+[tool.poetry.dev-dependencies]
+pytest = "^6.2"
+pytest-asyncio = "^0.16"
+python-dateutil = "^2.8"
```
Comment (pedropregueiro) [inline pyproject.toml:38]: last nitpick, this one is a dependency to `arrow`. guess it should also be on `.dependencies`?

Comment (kalaspuff) [reply pyproject.toml]: Indeed @pedropregueiro. Corrected that.

**Review by pedropregueiro — APPROVED** [review]
super!

**Comment by krellol** [comment]
omg posting here for time capsule reasons

hi mom!

**Pull request merged**